### PR TITLE
Parallelize perturbed with ThreadsX

### PR DESCRIPTION
**Pull request by BatyLeo**
Add an option to created  parallelized version of `PerturbedAdditive` and `PerturbedMultiplicative`, as a keyword argument `is_parallel` in their respective constructors.

See #29.

### CODE REVIEW

File: src/perturbed/abstract_perturbed.jl [hunk 1]
```diff
@@ -1,7 +1,8 @@
 """
-    AbstractPerturbed
+    AbstractPerturbed{B}
```
Comment (gdalle) [inline src/perturbed/abstract_perturbed.jl:2]: I would give a more explicit name to `B`, like `parallel`

Comment (gdalle) [reply src/perturbed/abstract_perturbed.jl:2]: Or at least `P` if you're too lazy

File: src/perturbed/additive.jl [hunk 2]
```diff
@@ -7,12 +7,19 @@ See also: [`AbstractPerturbed`](@ref).
...
     maximizer::F
     ε::Float64
     nb_samples::Int
     rng::R
     seed::S
+
+    function PerturbedAdditive{F,R,S,B}(
+        maximizer::F, ε::Float64, nb_samples::Int, rng::R, seed::S
+    ) where {F,R<:AbstractRNG,S<:Union{Nothing,Int},B}
+        @assert B isa Bool
```
Comment (gdalle) [inline src/perturbed/additive.jl:20]: I was also wondering how to constrain the type of a "value" type parameter, if you have any clues I'll take them!

File: src/InferOpt.jl [hunk 3]
```diff
@@ -14,6 +14,7 @@ using SparseArrays
 using Statistics
 using StatsBase: StatsBase, sample
 using Test
+using ThreadsX
```
Comment (gdalle) [inline src/InferOpt.jl:17]: Julia 1.8 lets you check how much time package loading takes, have you tried it to see if ThreadsX changes anything?

File: src/perturbed/multiplicative.jl [hunk 4]
```diff
@@ -28,12 +36,22 @@ end
...
 function PerturbedMultiplicative(
-    maximizer; ε=1.0, epsilon=nothing, nb_samples=1, rng=MersenneTwister(0), seed=nothing
-)
+    maximizer::F;
+    ε=1.0,
+    epsilon=nothing,
+    nb_samples=1,
+    rng::R=MersenneTwister(0),
+    seed::S=nothing,
+    is_parallel=false,
```
Comment (gdalle) [inline src/perturbed/multiplicative.jl:45]: Beware, this constructor is type unstable. Does it affect efficiency in the tests?

Comment (gdalle) [reply src/perturbed/multiplicative.jl:45]: To be clear, I don't think it matters, but I'd still be curious

File: src/perturbed/abstract_perturbed.jl (hunk 1 above)
Comment (BatyLeo) [reply src/perturbed/abstract_perturbed.jl:2]: I like parallel here

File: src/perturbed/additive.jl (hunk 2 above)
Comment (BatyLeo) [reply src/perturbed/additive.jl]: This is the only solution I found: https://discourse.julialang.org/t/restrict-type-of-value-type-parameter/12127

File: src/InferOpt.jl (hunk 3 above)
Comment (BatyLeo) [reply src/InferOpt.jl:17]: Do you have a link on how to do that ?

File: src/perturbed/multiplicative.jl (hunk 4 above)
Comment (BatyLeo) [reply src/perturbed/multiplicative.jl:45]: It doesn't seem to have changed tests duration

File: src/InferOpt.jl (hunk 3 above)
Comment (gdalle) [reply src/InferOpt.jl:17]: https://docs.julialang.org/en/v1/NEWS/#InteractiveUtils

File: src/perturbed/multiplicative.jl (hunk 4 above)
Comment (gdalle) [reply src/perturbed/multiplicative.jl:45]: how about profiling?

File: src/InferOpt.jl (hunk 3 above)
Comment (BatyLeo) [reply src/InferOpt.jl:17]: If I correctly interpreted the log, `ThreadsX` adds up only 20ms of compile time:

```julia
julia> @time_imports using InferOpt
[ Info: Precompiling InferOpt [4846b161-c94e-4150-8dac-c7ae193c601f]
      0.2 ms  Compat
     63.8 ms  ChainRulesCore
     20.1 ms  GenericSchur
      2.9 ms  ProgressMeter
      0.2 ms  ExprTools
     37.0 ms  TimerOutputs
     17.1 ms  MacroTools
      1.7 ms  StaticArraysCore
      0.4 ms  ConstructionBase
     20.3 ms  Setfield
    226.0 ms  MutableArithmetics
     60.7 ms  Parsers 8.15% compilation time
     11.6 ms  JSON
      1.8 ms  BenchmarkTools
      3.2 ms  IrrationalConstants
      0.4 ms  DiffRules
      1.4 ms  DiffResults
    394.3 ms  StaticArrays
      0.1 ms  OpenLibm_jll
      0.2 ms  NaNMath
      2.0 ms  DocStringExtensions 52.06% compilation time
      0.5 ms  ChangesOfVariables
      0.8 ms  InverseFunctions
      0.5 ms  LogExpFunctions
      0.2 ms  CompilerSupportLibraries_jll
      0.3 ms  OpenSpecFun_jll
     11.0 ms  SpecialFunctions
      0.2 ms  CommonSubexpressions
     80.6 ms  ForwardDiff
     49.2 ms  DataStructures
      1.3 ms  TranscodingStreams
      0.3 ms  Bzip2_jll
      0.5 ms  CodecBzip2
      0.1 ms  Zlib_jll
      0.8 ms  CodecZlib
   1285.7 ms  MathOptInterface
      0.1 ms  Hungarian
      2.2 ms  OpenBLAS_jll 94.37% compilation time
      0.3 ms  Arpack_jll
      5.6 ms  Arpack
    580.6 ms  FrankWolfe
      9.2 ms  Krylov
      0.1 ms  FastClosures
     22.5 ms  AMD 93.99% compilation time
      4.0 ms  LDLFactorizations
     28.0 ms  LinearOperators
      1.2 ms  SimpleTraits
      0.9 ms  DataAPI
      0.3 ms  SortingAlgorithms
      7.5 ms  Missings
      0.2 ms  StatsAPI
     19.2 ms  StatsBase
      8.2 ms  SplittablesBase
      0.4 ms  ArgCheck
     23.9 ms  InitialValues
      0.2 ms  Requires
      0.1 ms  DataValueInterfaces
      0.1 ms  IteratorInterfaceExtensions
      0.0 ms  TableTraits
     16.2 ms  Tables
      0.2 ms  ZygoteRules
    529.1 ms  BangBang 95.37% compilation time (71% recompilation)
      8.1 ms  MicroCollections
      0.2 ms  Adapt
      2.2 ms  Referenceables
     59.1 ms  Baselet
      0.1 ms  CompositionsBase
      0.1 ms  DefineSingletons
     54.4 ms  Transducers 42.80% compilation time
     18.6 ms  ThreadsX
      3.9 ms  InferOpt

```

Comment (gdalle) [reply src/InferOpt.jl:17]: Amazing!

File: src/perturbed/multiplicative.jl (hunk 4 above)
Comment (gdalle) [reply src/perturbed/multiplicative.jl:45]: Not sure this is necessary if the flame graph remains blue. Let's keep it simple

**Pull request merged**